File: exercises/advent2021/src/day2/part2.clj
(ns day2.part1
  (:require
   [clojure.string :as s]))

(def input-file
  (str "data/day2.txt"))

(defn parse-line
  [line]
  (let [[move amount] (s/split line #" ")]
    [move (Integer/parseInt amount)]))

(def input
  (->> input-file
       (slurp)
       (s/split-lines)
       (map parse-line)))

(defn step
  [[position depth aim]
   [move amount]]
  (case move
    "forward" [(+ position amount) (+ depth (* aim amount)) aim]
    "down" [position depth (+ aim amount)]
    "up" [position depth (- aim amount)]))

(def result
  (reduce step [0 0 0] input))

(def result
  (* (first result) (second result)))

(println result)

File: exercises/advent2021/src/day5/part2.py
import common

def solve(input: list[common.Line], size: int) -> int:
    board = [ [0]*size for i in range(size)]

    for line in input:
        dirX, dirY, length = 0, 0, 0

        if line.a.x < line.b.x:
            dirX = 1
            length = line.b.x - line.a.x
        elif line.a.x > line.b.x:
            dirX = -1
            length = line.a.x - line.b.x

        if line.a.y < line.b.y:
            dirY = 1
            length = line.b.y - line.a.y
        elif line.a.y > line.b.y:
            dirY = -1
            length = line.a.y - line.b.y

        for i in range(length + 1):
            board[line.a.y + (i*dirY)][line.a.x + (i*dirX)] += 1


    dangerPoints = 0        
    for y in range(size):
        for x in range(size):
            if board[y][x] > 1:
                dangerPoints += 1

    return dangerPoints

def main():
    input = common.parse('data/day5.txt')
    result = solve(input, 1000)
    print(result)

if __name__ == '__main__':
    main()